source/common/http/header_utility.cc

```diff
@@ -139,7 +139,10 @@ bool HeaderUtility::matchHeaders(const HeaderMap& request_headers, const HeaderD
 
   if (!header_value.result().has_value()) {
     if (header_data.invert_match_) {
-      return header_data.header_match_type_ == HeaderMatchType::Present && header_data.present_;
+      // For invert match with any match type but present, it should be considered as a match when
+      // the header is not present
+      return header_data.header_match_type_ == HeaderMatchType::Present
+                 ? header_data.present_ : true;
```

Comment:
I guess we need runtime flag to protect this.

And it will be great to update the document to describe the behavior
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-headermatcher

We can update the `invert_match_` example, to clarify this case.
```
invert_match
([bool](https://developers.google.com/protocol-buffers/docs/proto#scalar)) If specified, the match result will be inverted before checking. Defaults to false.

Examples:

The regex \d{3} does not match the value 1234, so it will match when inverted.

The range [-10,0) will match the value -1, so it will not match when inverted.
```